runtime/type-variable.js

```diff
@@ -66,42 +129,84 @@ class TypeVariable {
   set resolution(value) {
     assert(value instanceof Type);
     assert(!this._resolution);
+    let probe = value;
+    while (probe) {
+      if (!probe.isVariable)
+        break;
+      if (probe.variable == this)
+        return;
+      probe = probe.resolution;
+    }
+
     this._resolution = value;
-    this._constraint = null;
+    this._canWriteSuperset = null;
+    this._canReadSubset = null;
   }
 
-  get constraint() {
+  get canWriteSuperset() {
     if (this._resolution) {
-      assert(!this._constraint);
+      assert(!this._canWriteSuperset);
       if (this._resolution.isVariable) {
-        return this._resolution.variable.constraint;
+        return this._resolution.variable.canWriteSuperset;
       }
       return null;
     }
-    return this._constraint;
+    return this._canWriteSuperset;
   }
 
-  set constraint(value) {
+  set canWriteSuperset(value) {
     assert(!this._resolution);
-    this._constraint = value;
+    this._canWriteSuperset = value;
+  }
+
+  get canReadSubset() {
+    if (this._resolution) {
+      assert(!this._canReadSubset);
+      if (this._resolution.isVariable) {
+        return this._resolution.variable.canReadSubset;
+      }
+      return null;
+    }
+    return this._canReadSubset;
+  }
+
+  set canReadSubset(value) {
+    assert(!this._resolution);
+    this._canReadSubset = value;
+  }
+
+  maybeEnsureResolved() {
```

Comment:
maybe :p `tryForceResolution`?

Reply:
I just added canEnsureResolved .. let's chat about names when you get in